.github/workflows/main.yml

```
@@ -1,41 +1,54 @@
     defaults:
       run:
         shell: bash
     steps:
       - name: Checkout code
         uses: actions/checkout@v3
-      - name: gofmt
+      - name: Set up Go environment
+        uses: actions/setup-go@v3
+        with:
+          go-version: '1.18.8'
+          cache: true
+      - name: Check Go and tidy cache
+        run: |
+          echo "* Golang version: $(go version)"
+          echo "* HOME value in this node: $HOME"
+          echo "* Running go mod tidy"
+          go mod tidy
+      - name: Run gofmt
         # Run gofmt first, as it's quick and issues are common.
         run: diff -u <(echo -n) <(gofmt -s -d $(git ls-files '*.go'))
-      - name: go test
+      - name: Run go test cover
         run: |
           # check that log lines contains no invalid chars (evidence of format mismatch)
           export LOG_PANIC_ON_INVALIDCHARS=true
           # we run vet in another step
           go test -vet=off -timeout=1m -coverprofile=covprofile ./...
+      - name: Run go test race
+        run: |
+          # check that log lines contains no invalid chars (evidence of format mismatch)
+          export LOG_PANIC_ON_INVALIDCHARS=true
           # -race can easily make the crypto stuff 10x slower
           go test -vet=off -timeout=15m -race ./...
-      - name: go vet
+      - name: Run go vet
         run: go vet ./...
       - name: Download staticcheck
+        # staticcheck provides a github action, use it: https://staticcheck.io/docs/running-staticcheck/ci/github-actions/
         run: |
           curl -L https://github.com/dominikh/go-tools/releases/download/v0.3.3/staticcheck_linux_amd64.tar.gz | tar -xzf -
-      - name: staticcheck
+      - name: Run staticcheck
         run: |
           ./staticcheck/staticcheck ./... 2> staticcheck/stderr
-      - name:  check staticcheck stderr
+      - name: Check staticcheck stderr (this step isn't needed because we are using actions/setup-go@v3 on GitHub hosted runner)
```

This is all done manually due to a bug in Go, where bad permissions can lead to staticcheck silently doing nothing. See https://github.com/golang/go/issues/50869.

We can switch to just running staticcheck normally or via its github action, but then we risk silently not running staticcheck again in the future. IMO we should leave this workaround in place, even if it's clunky, until we use a Go version with that bugfix included.

---

Since it is a limitation of Go, using another Linter or StaticCheck is not going to fix the problem.
Fair enough!